README.md
# govee-lan-server
Node.js web server for controlling Govee lighting devices via their local LAN API   

I've also written a [Python version called govee-lan-server-py](https://github.com/20goto10/govee-lan-server-py) that operates exactly the same. That one is more stable, but this one only does device discovery once (making it a bit faster-- too bad it is currently defective).


# What it does
Govee-lan-server is a lightweight web server that can handle a few simple actions for controlling one or more Govee lights. This uses their new local LAN API, via the [Govee-lan-control](https://github.com/Joery-M/Govee-LAN-Control) Node library. It does NOT use Govee's remote API or BLE; it must run within the device's LAN.

All of this could be done in standalone scripts (per the examples in [Govee-lan-control](https://github.com/Joery-M/Govee-LAN-Control)'s documentation), but I wanted an always-on event listener to reduce latency caused by the startup "discovery" phase. Using a web server is the easiest way to make it reachable from the separate system that will be controlling it (a rooted Philips Hue hub). I use a Philips Hue dimmer switch control my Govee lights, using a Python script I wrote that runs on a rooted Hue: [hue-jazz](https://github.com/20goto10/hue-jazz/). You could also issue the same requests from other home automation software. 

I will likely abandon this project as soon as there is a good OpenHAB binding for these kinds of light strips.

# Warning!
Don't run this on a publicly accessible server, unless you know what you're doing. There's no access control, and I don't want to be responsible for any involuntary lightshows. The risk of a serious hack is pretty small but I'd rather not be responsible for any part of it.

# How to do it

0. You'll need a modern version of Node.js. I'm on 16 but older versions would probably work. Assuming that's taken care of--
```
git clone https://github.com/20goto10/govee-lan-server
cd govee-lan-server
npm install
cp config.json.sample config.json
```


1. Make sure you have the Govee "LAN Control" option in Govee's confusing app. This is not a straightforward matter. I had to try several things to wake this up. I'm not sure what eventually got it to work, but it might be the tip I found on Reddit to link the device to your phone's wi-fi hotspot. I also deleted and reset the device, power-cycled it numerous times, and restarted the app numerous times. All I can say for sure is that I have an H61E1 and it did eventually work. Why the LAN Control option is not visibile by default, or easy to enable, is anyone's guess. If you unplug it and plug it back in 30 minutes later as the Govee docs suggest, you have probably just wasted 30 minutes.

2. Edit `config.json` and map device IDs to whatever device nicknames you wish to use. If you're not sure of the device ID, just launch the server (per the next step). When the server discovers a device, it will print out the ID. Use that as a key in the 'device_map' section of the JSON config, with any nickname you like for the value. 

   The rest of the config, currently, is just a port setting for your server (default should be fine) and a default_fade_time if you want to fade colors over time (based on the examples from [Govee-lan-control](https://github.com/Joery-M/Govee-LAN-Control)), which can be left as-is. 

   (I realize the device nicknames aren't really necessary but I prefer having a human-readable identifier in the URL so I understand my automations later. The config also makes it possible to avoid providing control of any devices you want to omit, for whatever reason may someday arise. I only have one Govee device in service so in my case it's a moot point.)

3. Once you have your config map set up, start the app.
```
node govee-lan-server.js
````


4. Once the app is set up, you can trigger the HTTP requests however you like (e.g. from curl, or a browser). See below for examples.

5. (optional) A systemd service daemon script is included in the `service` directory. It must be updated and copied to the correct location. See README in that directory.


## Using the API

Note that if running these curl commands in a terminal you will need to escape the & characters (i.e. change them to "\\&").

To run them via a browser you can just omit the word "curl". (Well, if you made it this far, you probably know that, but anyway...) The server doesn't care what HTTP request method you use.

Supposing it's running at 192.168.1.230, on the default port of 3666, and your device nickname in the device_map is "lightstrip".

To turn it on:

`curl http://192.168.1.230:3666?device=lightstrip&action=on`

To turn it off:

`curl http://192.168.1.230:3666?device=lightstrip&action=off`

Set the brightness (in this case to 50%). The brightness scale is 0-100.

`curl http://192.168.1.230:3666?device=lightstrip&action=brightness&value=50`

Raise or lower the brightness relatively up or down by a given amount with the "delta" param: reads the brightness value and adjusts accordingly, by percentage (of the maximum). Use negative numbers to decrease. Here we lower it by 5%:

`curl http://192.168.1.230:3666?device=lightstrip&action=brightness&delta=-5`

Set the color, in this case, to 100% green-ness. 

`curl http://192.168.1.230:3666?device=lightstrip&action=color&value=00FF00`

Fade the color, in this case, to pale blue, with a 3 second fade time (note that if "fade" is not supplied, it will use the default from config).

`curl http://192.168.1.230:3666?device=lightstrip&action=color&value=DDFFDD&fade=3000`


That's pretty much it.  Valid actions are:
```
on
off
color
brightness
fadecolor
```


# Contributing
Please, feel free. My apologies for any sloppy code or Node offenses-- this is something I hacked together in a couple hours in vi on a Raspberry pi.

# Road map

- *delta fix* -- brightening/dimming via the delta option doesn't really work the way people would want to use it, because the Govee doesn't update its state fast enough. So if you request dimming several times, you will probably get the same result as doing it once. There may be a way to trigger a state update, but most likely the solution is to not count on the device's brightness setting except the first time, and maintain it as independent state (e.g. in this app). Then set the brightness to a pre-calculated value based on that, and not by using the current (inaccurate) state.

- *RGBIC control?*
If there is some way of taking advantage of the RGBIC features I'd love to know about it. It's hardly urgent, and I won't be the one to figure that out. Honestly I don't really care about the animations-- I use the strip like track lighting, and can use their insanely complex app when I feel like setting a Halloween theme or something.

- *de-sillification*
There's some half-baked stuff in here. For example, there's no real need for an "action" parameter at all. The whole operation could be done by the request payload. Nicknames are pointless. The config is case-sensitive about IDs. This allows GET requests to do things that should require POST requests. Etc. All of this is true, but it's not meant to be an enterprise solution, just a quick fix to the problem. Still, eventually I might de-sillify it.

# Thank you
I did the easy work, not the hard work. The hard work was done by [Joery-M](https://github.com/Joery-M), namely their [Govee-lan-control](https://github.com/Joery-M/Govee-LAN-Control) project. Many thanks, Joery-M!

Thank you to Govee for finally adding this LAN control feature. This has the side-effect of taking a cheap light strip and turning it into a cheap good one.

# Ranting
All IoT devices should have a local API. The trend of having them handling their most basic jobs over the internet, instead of strictly within the LAN, has always struck me as insane. 



CHANGELOG.md
# v0.01 / 2023-01-30
- change setOn and setOff to setColor controls. On/Off stopped working for me -- is it possible a firmware update changed the API?

# v0.009 / 2022-12-26
- Initial creation of everything



govee-test.js
// Simple test script based on the govee-lan-control examples (which are buggy)
//
const Govee = require("govee-lan-control");

var govee = new Govee.default();
govee.on("ready", () => {
  console.log("Server/client is ready!");
});
govee.on("deviceAdded", (device) => {
  console.log("New Device!", device.model);
  device.actions.setOn();

  setInterval(() => {
    device.actions.fadeColor({
      time: 2000,
      color: {
        hex: Math.floor(Math.random() * 16777215).toString(16),
      },
      brightness: Math.random() * 100,
    });
  }, 2000);
});





server.js
// To fix: 
//   Startup issues (infrequent):
//   - sometimes the server seems to hang at startup.
//   - sometimes the device doesn't show up to register itself, for no known reason. 

const http = require("http");
const fs = require('fs');
const url = require('url');
const Govee = require("govee-lan-control");

var config = { port: 3666, // default, can be overridden in config
               default_fade_time: 1000 } // for fadecolor; can be overridden in URL
const config_file_name = 'config.json';
const VALID_ACTIONS = ['on', 'off', 'brightness', 'color', 'fadecolor'];

var govee = new Govee.default();
var devices = {};

const handle_action = (govee_device, nickname, action, params) => {
  switch(action) { 
    case 'on': govee_device.actions.setOn();
	       return { message: "turned on " + nickname + " (" + govee_device.deviceID + ")" };
	       break;
    case 'off': govee_device.actions.setOff();
	        return { message: "turned off " + nickname + " (" + govee_device.deviceID + ")" };
	        break;
    case 'brightness': if (params['value']) {
	                 govee_device.actions.setBrightness(params.value);
	                 return { message: "set " + nickname + " brightness to " + params.value + " (" + govee_device.deviceID + ")" };
	               }
	               else {
			 return { error: "brightness needs a 'value' setting" };
		       }
	               break;
    case 'color': if (params['value']) {
	            govee_device.actions.setColor({ hex: params.value });
	            return { message: "set " + nickname + " color to " + params.value + " (" + govee_device.deviceID + ")" };
	          }
	          else {
	            return { error: "color needs a 'value' setting (a hex RGB string prefixed with #, like '#abc123')" };
		  }
		  break;
    case 'fadecolor': if (params['value']) {
	            var fade_time = config['default_fade_time'];
	            var param_note = '(config default_fade_time) ';
	            if (params['time']) {
	              fade_time = params.time;
		      param_note = "(time from request) ";
                    }
	            var options = { color: { hex: params.value }, 
		                    time: fade_time };
	            if (params['brightness']) {
		      // It doesn't really make sense to have a brightness
		      // AND a hex code for the color, but, the option is 
		      // there in the lib, so who am I to say you can't do it.

	              options['brightness'] = params.brightness;
		      param_note += "brightness " + params.brightness;
                    }
	            govee_device.actions.fadeColor(options);
	            return { message: "fading " + nickname + " color to " + params.value + " over " + fade_time + "msec " + param_note + " (" + govee_device.deviceID + ")" };
	          }
	          else {
	            return { error: "color needs a 'value' setting (a hex RGB string prefixed with #, like '#abc123')" };
		  }
		  break;
  }
  return { error: "unknown" };
};

const parse_request = (queryObject) => {
  if (queryObject.device) { 
    if (devices[queryObject.device]) {
      if (queryObject.action) {
	if (VALID_ACTIONS.indexOf(queryObject.action) != -1) { 
          return handle_action(devices[queryObject.device], queryObject.device, queryObject.action, queryObject);
	}
	else {
	  return { error: "requested action " + queryObject.action + " is not a valid action; valid actions are: " + VALID_ACTIONS.join(',') };
	}
      } else {
	return { error: "no action in URL query" };
      }
    } else {
      return { error: "device " + queryObject.device + " is not known (yet?); make sure it has a value in the config.json 'device_map' object" };
    }
  } else {
    return { error: "no device in URL query" };
  }
  return {};
};

const server = http.createServer((req, res) => {
  const queryObject = url.parse(req.url, true).query;
  parsed_request = parse_request(queryObject);
  if (parsed_request['error']) {
    res.writeHead(400, {'Content-Type': 'application/json'});
    res.write(JSON.stringify(parsed_request));
  } else {
    res.writeHead(200, {'Content-Type': 'application/json'});
    res.write(JSON.stringify(parsed_request));
  }
  res.end();
});

const start_server = () => {
  fs.readFile(config_file_name, 'utf8', (err, data) => {
    if (err) {
      console.log(`Error reading file from disk: ${err}`)
    } else {
      // merge existing default config with whatever is in the file
      Object.assign(config, JSON.parse(data));
      // TODO: should really make the device keys case-insensitive...
	    
      console.log("Config: ", config);

      govee.on("deviceAdded", (device) => {
        console.log("Found new device:", device.deviceID, "ip address: ", device.ip, "model:", device.model);
        if (config['device_map'] && config['device_map'][device.deviceID]) {
          console.log("Device " + device.deviceID + " is mapped to name " + config['device_map'][device.deviceID]);
          devices[config['device_map'][device.deviceID]] = device;
	} else {
          console.log("Could not find device " + device.deviceID + " in config['device_map'] (this could be a matter of case sensitivity); ignoring device");
	}
      });
 
      server.listen((config['port']), () => {
        console.log("Server is running on Port " + config['port']);
      })
    }
  })
};

govee.on("ready", () => {
  console.log("Govee listener is ready; starting web server");
  start_server();
});



govee-lan-server.js
// To fix: 
//   Startup issues (infrequent):
//   - sometimes the server seems to hang at startup.
//   - sometimes the device doesn't show up to register itself, for no known reason. 

const http = require("http");
const fs = require('fs');
const url = require('url');
const Govee = require("govee-lan-control");

var config = { port: 3666, // default, can be overridden in config
               default_fade_time: 1000 } // for fadecolor; can be overridden in URL
const config_file_name = 'config.json';
const VALID_ACTIONS = ['on', 'off', 'brightness', 'color', 'fadecolor'];

var govee = new Govee.default();
var devices = {};

const handle_action = (govee_device, nickname, action, params) => {
  response = { error: "unknown" };
  switch(action) { 
	  case 'on': govee_device.actions.setColor({ hex: '#ffffff' }); // rgb: 255,255,255); //Brightness(255); // On();
	       response = { message: "turned on " + nickname + " (" + govee_device.deviceID + ")" };
	       break;
	  case 'off': govee_device.actions.setColor({ hex: '#000000' }); // setOff();
	        response = { message: "turned off " + nickname + " (" + govee_device.deviceID + ")" };
	        break;
    case 'brightness': if (params['value']) {
	                 govee_device.actions.setBrightness(params.value);
	                 response = { message: "set " + nickname + " brightness to " + params.value + " (" + govee_device.deviceID + ")" };
	               } 
		       else if (params['delta']) {
                         var current_brightness = govee_device.state.brightness;
			 var delta = parseInt(params['delta']);
                         var new_brightness = current_brightness + delta;
                         if (new_brightness > 100) {
				 new_brightness = 100;
			 } else if (new_brightness < 0) {
				 new_brightness = 0;
			 }
	                 govee_device.actions.setBrightness(new_brightness);
	                 response = { message: "adjust " + nickname + " brightness by " + delta + "%, from " + current_brightness + " to " + new_brightness + " (" + govee_device.deviceID + ")" };
		       }
	               else {
			 response = { error: "brightness needs a 'value' or 'delta' setting" };
		       }
	               break;
    case 'color': if (params['value']) {
	            govee_device.actions.setColor({ hex: '#' + params.value });
	            response = { message: "set " + nickname + " color to " + params.value + " (" + govee_device.deviceID + ")" };
	          }
	          else {
	            response = { error: "color needs a 'value' setting (a hex RGB string prefixed with #, like '#abc123')" };
		  }
		  break;
    case 'fadecolor': if (params['value']) {
	                var fade_time = config['default_fade_time'];
	                var param_note = '(config default_fade_time) ';
	                if (params['time']) {
	                  fade_time = params.time;
		          param_note = "(time from request) ";
                        }
	                var options = { color: { hex: params.value }, 
		                        time: fade_time };
	                if (params['brightness']) {
		          // It doesn't really make sense to have a brightness
		          // AND a hex code for the color, but, the option is 
		          // there in the lib, so who am I to say you can't do it.

	                 options['brightness'] = params.brightness;
		         param_note += "brightness " + params.brightness;
                        }
	                govee_device.actions.fadeColor(options);
	                response = { message: "fading " + nickname + " color to " + params.value + " over " + fade_time + "msec " + param_note + " (" + govee_device.deviceID + ")" };
	              } else {
	                response = { error: "color needs a 'value' setting (a hex RGB string prefixed with #, like '#abc123')" };
		      }
		      break;
  }
  return response;
};

const parse_request = (queryObject) => {
  if (queryObject.device) { 
    if (devices[queryObject.device]) {
      if (queryObject.action) {
	if (VALID_ACTIONS.indexOf(queryObject.action) != -1) { 
          return handle_action(devices[queryObject.device], queryObject.device, queryObject.action, queryObject);
	}
	else {
	  return { error: "requested action " + queryObject.action + " is not a valid action; valid actions are: " + VALID_ACTIONS.join(',') };
	}
      } else {
	return { error: "no action in URL query" };
      }
    } else {
      return { error: "device " + queryObject.device + " is not known (yet?); make sure it has a value in the config.json 'device_map' object" };
    }
  } else {
    return { error: "no device in URL query" };
  }
  return {};
};

const server = http.createServer((req, res) => {
  const queryObject = url.parse(req.url, true).query;
  parsed_request = parse_request(queryObject);
  if (parsed_request['error']) {
    res.writeHead(400, {'Content-Type': 'application/json'});
    res.write(JSON.stringify(parsed_request));
  } else {
    res.writeHead(200, {'Content-Type': 'application/json'});
    res.write(JSON.stringify(parsed_request));
  }
  res.end();
});

const start_server = () => {
  fs.readFile(config_file_name, 'utf8', (err, data) => {
    if (err) {
      console.log(`Error reading file from disk: ${err}`)
    } else {
      // merge existing default config with whatever is in the file
      Object.assign(config, JSON.parse(data));
      // TODO: should really make the device keys case-insensitive...
	    
      console.log("Config: ", config);

      govee.on("deviceAdded", (device) => {
        console.log("Found new device:", device.deviceID, "ip address: ", device.ip, "model:", device.model);
        // console.log("Full device map:", device);
        if (config['device_map'] && config['device_map'][device.deviceID]) {
          console.log("Device " + device.deviceID + " is mapped to name " + config['device_map'][device.deviceID]);
          devices[config['device_map'][device.deviceID]] = device;
	} else {
          console.log("Could not find device " + device.deviceID + " in config['device_map'] (this could be a matter of case sensitivity); ignoring device");
	}
      });
 
      server.listen((config['port']), () => {
        console.log("Server is running on Port " + config['port']);
      })
    }
  })
};

govee.on("ready", () => {
  console.log("Govee listener is ready; starting web server");
  start_server();
});

